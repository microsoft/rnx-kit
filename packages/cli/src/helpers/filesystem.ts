import * as nodefs from "fs"; // Cannot use `node:fs` because of Jest mocks

export function ensureDir(p: string, fs = nodefs): void {
  fs.mkdirSync(p, { recursive: true, mode: 0o755 });
}
