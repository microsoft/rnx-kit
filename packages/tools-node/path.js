module.exports = require("./lib/path");
