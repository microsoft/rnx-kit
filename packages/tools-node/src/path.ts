/**
 * Escape a path by replacing each backslash ('\\') with a double-backslash ("\\\\").
 *
 * @param p Path to escape
 * @returns Escaped path
 */
export function escapePath(p: string): string {
  return p.replace(/\\/g, "\\\\");
}

/**
 * Normalize the separators in a path, converting each backslash ('\\') to a foreward
 * slash ('/').
 *
 * @param p Path to normalize
 * @returns Normalized path
 */
export function normalizePath(p: string): string {
  return p.replace(/\\/g, "/");
}
