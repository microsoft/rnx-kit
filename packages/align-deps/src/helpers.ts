import type { PackageManifest } from "@rnx-kit/tools-node/package";
import { writePackage } from "@rnx-kit/tools-node/package";
import detectIndent from "detect-indent";
import fs from "fs";
import semverValidRange from "semver/ranges/valid";

export function compare<T>(lhs: T, rhs: T): -1 | 0 | 1 {
  if (lhs === rhs) {
    return 0;
  } else if (lhs < rhs) {
    return -1;
  } else {
    return 1;
  }
}

export function dropPatchFromVersion(version: string): string {
  return version
    .split("||")
    .map((input) => {
      const versionRange = input.trim();
      if (!semverValidRange(versionRange)) {
        throw new Error(`Invalid version number: ${versionRange}`);
      }

      if (!versionRange) {
        return "*";
      }

      return versionRange
        .split(" ")
        .map((v) => {
          if (v === "*" || v === "-") {
            // No need to manipulate `*` or hyphen ranges, e.g. `1.0 - 2.0`
            return v;
          }

          const [major, minor = "0"] = v.split(".");
          return major === "^0" || major === "~0"
            ? `0.${minor}`
            : `${major}.${minor}`;
        })
        .join(" ");
    })
    .join(" || ");
}

export function modifyManifest(
  pkgPath: string,
  manifest: PackageManifest
): void {
  const content = fs.readFileSync(pkgPath, { encoding: "utf-8" });
  const indent = detectIndent(content).indent || "  ";
  writePackage(pkgPath, manifest, indent);
}

export function omitEmptySections(manifest: PackageManifest): PackageManifest {
  const sections = ["dependencies", "peerDependencies", "devDependencies"];
  for (const sectionName of sections) {
    const section = manifest[sectionName];
    if (typeof section === "object" && Object.keys(section).length === 0) {
      delete manifest[sectionName];
    }
  }
  return manifest;
}
