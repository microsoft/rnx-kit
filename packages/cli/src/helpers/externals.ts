import { resolveDependencyChain } from "@rnx-kit/tools-node/package";
import { resolveCommunityCLI } from "@rnx-kit/tools-react-native/context";
import type { CliServerApi, CoreDevMiddleware } from "../serve/types";

type ExternalModule =
  | "@react-native-community/cli-server-api"
  | "@react-native/dev-middleware";

function friendlyRequire<T>(modules: string[], startDir: string): T {
  try {
    const modulePath = resolveDependencyChain(modules, startDir);
    return require(modulePath) as T;
  } catch (_) {
    const module = modules[modules.length - 1];
    throw new Error(
      `Cannot find module '${module}'. This probably means that ` +
        "'@rnx-kit/cli' is not compatible with the version of 'react-native' " +
        "that you are currently using. Please update to the latest version " +
        "and try again. If the issue still persists after the update, please " +
        "file a bug at https://github.com/microsoft/rnx-kit/issues."
    );
  }
}

export function requireExternal(
  module: "@react-native-community/cli-server-api",
  projectRoot: string,
  reactNativePath: string
): CliServerApi;

export function requireExternal(
  module: "@react-native/dev-middleware",
  projectRoot: string,
  reactNativePath: string
): CoreDevMiddleware;

export function requireExternal(
  module: ExternalModule,
  projectRoot: string,
  reactNativePath: string
): CliServerApi | CoreDevMiddleware {
  switch (module) {
    case "@react-native-community/cli-server-api":
      return friendlyRequire<CliServerApi>(
        [
          "@react-native-community/cli",
          "@react-native-community/cli-server-api",
        ],
        resolveCommunityCLI(projectRoot, reactNativePath)
      );
    case "@react-native/dev-middleware":
      return friendlyRequire<CoreDevMiddleware>(
        ["@react-native/community-cli-plugin", "@react-native/dev-middleware"],
        reactNativePath
      );
  }
}
