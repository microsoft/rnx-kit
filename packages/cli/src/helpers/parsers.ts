import { InvalidArgumentError } from "commander";
import type { TransformProfile } from "metro-babel-transformer";
import * as path from "node:path";

export function asBoolean(value: string): boolean {
  switch (value) {
    case "false":
      return false;
    case "true":
      return true;
    default:
      throw new InvalidArgumentError(`Expected 'true' or 'false'.`);
  }
}

export function asNumber(value: string): number {
  return Number(value);
}

export function asResolvedPath(value: string): string {
  return path.resolve(value);
}

export function asStringArray(value: string): string[] {
  return value.split(",");
}

export function asTransformProfile(val: string): TransformProfile {
  switch (val) {
    case "hermes-stable":
    case "hermes-canary":
    case "default":
      return val;

    default: {
      const profiles: TransformProfile[] = [
        "hermes-stable",
        "hermes-canary",
        "default",
      ];
      throw new InvalidArgumentError(`Expected '${profiles.join("', '")}'.`);
    }
  }
}
