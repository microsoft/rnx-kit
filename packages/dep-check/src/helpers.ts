import { warn } from "@rnx-kit/console";
import type { PackageManifest } from "@rnx-kit/tools-node/package";
import { writePackage } from "@rnx-kit/tools-node/package";
import detectIndent from "detect-indent";
import fs from "fs";

export function compare<T>(lhs: T, rhs: T): -1 | 0 | 1 {
  if (lhs === rhs) {
    return 0;
  } else if (lhs < rhs) {
    return -1;
  } else {
    return 1;
  }
}

export function concatVersionRanges(versions: string[]): string {
  return "^" + versions.join(" || ^");
}

export function keysOf<T extends Record<string, unknown>>(obj: T): (keyof T)[] {
  return Object.keys(obj);
}

export function modifyManifest(
  pkgPath: string,
  manifest: PackageManifest
): void {
  const content = fs.readFileSync(pkgPath, { encoding: "utf-8" });
  const indent = detectIndent(content).indent || "  ";
  writePackage(pkgPath, manifest, indent);
}

export function omitEmptySections(manifest: PackageManifest): PackageManifest {
  const sections = ["dependencies", "peerDependencies", "devDependencies"];
  for (const sectionName of sections) {
    const section = manifest[sectionName];
    if (typeof section === "object" && Object.keys(section).length === 0) {
      delete manifest[sectionName];
    }
  }
  return manifest;
}

export function printMigrationMessage(): void {
  const banner =
    "⚠️   ⚠️   ⚠️   ⚠️   ⚠️   ⚠️   ⚠️   ⚠️   ⚠️   ⚠️   ⚠️   ⚠️   ⚠️   ⚠️   ⚠️   ⚠️";

  warn(banner);
  warn("'@rnx-kit/dep-check' has been renamed to '@rnx-kit/align-deps'!");
  warn(
    "You can replace '@rnx-kit/dep-check' with '@rnx-kit/align-deps' in your 'package.json' and your configurations will be automatically upgraded."
  );
  warn(
    "For reasoning and more details, you can read the RFC: https://github.com/microsoft/rnx-kit/pull/1757"
  );
  warn(banner);
}
