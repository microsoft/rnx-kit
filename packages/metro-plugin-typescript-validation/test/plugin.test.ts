import type {
  Dependency,
  Graph,
  Module,
  MixedOutput,
  SerializerOptions,
} from "@rnx-kit/metro-serializer";
import {
  getModuleRoot,
  readTsConfig,
  writeMetroTsConfigToNodeModules,
  runTypeScriptCompiler,
  visit,
  TypeScriptValidation,
} from "../src/plugin";
import path from "path";
const child_process = require("child_process");
const fs = require("fs");
const yargs = require("yargs/yargs");
import { hideBin } from "yargs/helpers";

jest.mock("child_process");
jest.mock("fs");
jest.mock("yargs/yargs");
jest.mock("yargs/helpers");

//  test data

const rootPathPosix = "/repos/yoyodyne/packages/overthruster";
const rootPathWin32 = path.win32.normalize("C:" + rootPathPosix);

const graphPosix: Graph = {
  dependencies: new Map<string, Module<MixedOutput>>(),
  importBundleNames: new Set(),
  entryPoints: [rootPathPosix + "/src/main.ts"],
};
const graphWin32: Graph = {
  dependencies: new Map<string, Module<MixedOutput>>(),
  importBundleNames: new Set(),
  entryPoints: [rootPathWin32 + "\\src\\main.ts"],
};

function addGraphDependency(g: Graph, modulePath: string) {
  g.dependencies.set(modulePath, {
    dependencies: new Map<string, Dependency>(),
    inverseDependencies: new Set<string>(),
    output: [],
    path: modulePath,
    getSource: () => {
      return null;
    },
  });
  return g.dependencies.get(modulePath);
}

function addModuleDependency(
  g: Graph,
  parent,
  moduleName: string,
  modulePath: string
) {
  parent.dependencies.set(moduleName, {
    absolutePath: modulePath,
  });

  return addGraphDependency(g, modulePath);
}

const index_ts_posix = addGraphDependency(
  graphPosix,
  rootPathPosix + "/src/main.ts"
);
const index_ts_win32 = addGraphDependency(
  graphWin32,
  rootPathWin32 + "\\src\\main.ts"
);

const propulsion_ts_posix = addModuleDependency(
  graphPosix,
  index_ts_posix,
  "./src/propulsion.ts",
  rootPathPosix + "/src/propulsion.ts"
);
const propulsion_ts_win32 = addModuleDependency(
  graphWin32,
  index_ts_win32,
  ".\\src\\propulsion.ts",
  rootPathWin32 + "\\src\\propulsion.ts"
);

const dimensions_ts_posix = addModuleDependency(
  graphPosix,
  index_ts_posix,
  "./src/dimensions.ts",
  rootPathPosix + "/src/dimensions.ts"
);
const dimensions_ts_win32 = addModuleDependency(
  graphWin32,
  index_ts_win32,
  ".\\src\\dimensions.ts",
  rootPathWin32 + "\\src\\dimensions.ts"
);

const lectroid_js_posix = addModuleDependency(
  graphPosix,
  index_ts_posix,
  "./src/lectroid.js",
  rootPathPosix + "/src/lectroid.js"
);
const lectroid_js_win32 = addModuleDependency(
  graphWin32,
  index_ts_win32,
  ".\\src\\lectroid.js",
  rootPathWin32 + "\\src\\lectroid.js"
);

const react_native_posix = addModuleDependency(
  graphPosix,
  index_ts_posix,
  "react-native",
  "/repos/yoyodyne/node_modules/react-native/index.js"
);
const react_native_win32 = addModuleDependency(
  graphWin32,
  index_ts_win32,
  "react-native",
  "C:\\repos\\yoyodyne\\node_modules\\react-native\\index.js"
);

//  create a circular dependency
propulsion_ts_posix.dependencies.set("./src/dimensions.ts", {
  absolutePath: rootPathPosix + "/src/dimensions.ts",
  data: null,
});
propulsion_ts_win32.dependencies.set(".\\src\\dimensions.ts", {
  absolutePath: rootPathWin32 + "\\src\\dimensions.ts",
  data: null,
});

dimensions_ts_posix.dependencies.set("./src/propulsion.ts", {
  absolutePath: rootPathPosix + "/src/propulsion.ts",
  data: null,
});
dimensions_ts_win32.dependencies.set(".\\src\\propulsion.ts", {
  absolutePath: rootPathPosix + "\\src\\propulsion.ts",
  data: null,
});

//  test suite

describe("getModuleRoot()", () => {
  test("throws on unknown module", () => {
    expect(() => {
      getModuleRoot("not-a-real-module");
    }).toThrowError();
  });

  test("resolves to the root of @msfast/typescript-platform-resolution", () => {
    expect(getModuleRoot("@msfast/typescript-platform-resolution")).toEqual(
      expect.stringMatching(/@msfast[/\\]typescript-platform-resolution$/)
    );
  });
});

describe("readTsConfig()", () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  test("reads the contents of tsconfig.json", () => {
    fs.readFileSync.mockReturnValue(
      JSON.stringify({
        compilerOptions: { noEmit: true },
        include: ["src"],
      })
    );

    const p = path.normalize("/path/to/file");
    expect(readTsConfig(p)).toMatchSnapshot();
    expect(fs.readFileSync).toBeCalledTimes(1);
    expect(fs.readFileSync).toBeCalledWith(
      path.join(p, "tsconfig.json"),
      expect.anything()
    );
  });
});

describe("writeMetroTsConfigToNodeModules()", () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  test("writes the contents of tsconfig to a temp file in node_modules", () => {
    const projectRoot = "/root/path/here";
    const tsconfig = { a: 123 };

    writeMetroTsConfigToNodeModules(projectRoot, tsconfig);

    expect(fs.writeFileSync).toBeCalledTimes(1);
    expect(fs.writeFileSync).toBeCalledWith(
      expect.stringMatching(/[/\\]node_modules[/\\]tsconfig-metro.json$/),
      '{"a":123}'
    );
  });
});

describe("runTypeScriptCompiler()", () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  test("executes the typescript compiler using node.js", () => {
    child_process.spawnSync.mockReturnValue({});

    runTypeScriptCompiler("/path/to/project/tsconfig.json");

    expect(child_process.spawnSync).toBeCalledTimes(1);
    const spawnSyncParams = child_process.spawnSync.mock.calls[0];
    const { [0]: executable, [1]: args } = spawnSyncParams;
    expect(executable).toEqual(process.execPath);
    expect(args[0]).toEqual(
      expect.stringMatching(
        /[/\\]@msfast[/\\]typescript-platform-resolution.*[/\\]tsc.js$/
      )
    );
  });

  test("succeeds when the process exit code is zero", () => {
    child_process.spawnSync.mockReturnValue({
      status: 0,
    });

    runTypeScriptCompiler("/path/to/project/tsconfig.json");
    expect(child_process.spawnSync).toBeCalledTimes(1);
  });

  test("throws when the process exit code is non-zero", () => {
    child_process.spawnSync.mockReturnValue({
      status: 1,
    });

    expect(() =>
      runTypeScriptCompiler("/path/to/project/tsconfig.json")
    ).toThrowErrorMatchingSnapshot();
  });

  test("throws when the process exit code is non-zero and an error is given", () => {
    child_process.spawnSync.mockReturnValue({
      status: 1,
      error: new Error("simulated error message from typescript compiler"),
    });

    expect(() =>
      runTypeScriptCompiler("/path/to/project/tsconfig.json")
    ).toThrowErrorMatchingSnapshot();
  });

  test("throws when the process crashes", () => {
    child_process.spawnSync.mockReturnValue({
      signal: "SIGKILL",
    });

    expect(() =>
      runTypeScriptCompiler("/path/to/project/tsconfig.json")
    ).toThrowErrorMatchingSnapshot();
  });

  test("throws when the process exit code is non-zero and an error is given", () => {
    child_process.spawnSync.mockReturnValue({
      signal: "SIGKILL",
      error: new Error(
        "simulated crash message caused by sending SIGKILL to the typescript compiler"
      ),
    });

    expect(() =>
      runTypeScriptCompiler("/path/to/project/tsconfig.json")
    ).toThrowErrorMatchingSnapshot();
  });
});

describe("visit()", () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  test("traverses the entire graph (posix)", () => {
    const visited: Record<string, boolean> = {};
    const files: Array<string> = [];

    graphPosix.entryPoints.forEach((m) =>
      visit(m, graphPosix, [rootPathPosix], visited, files)
    );
    expect(visited).toMatchSnapshot();
  });

  test("traverses the entire graph (win32)", () => {
    const visited: Record<string, boolean> = {};
    const files: Array<string> = [];

    graphWin32.entryPoints.forEach((m) =>
      visit(m, graphWin32, [rootPathWin32], visited, files)
    );
    expect(visited).toMatchSnapshot();
  });

  test("returns a list of files in the current package (posix)", () => {
    const visited: Record<string, boolean> = {};
    const files: Array<string> = [];

    graphPosix.entryPoints.forEach((m) =>
      visit(m, graphPosix, [rootPathPosix], visited, files)
    );
    expect(files).toMatchSnapshot();
  });

  test("returns a list of files in the current package (win32)", () => {
    const visited: Record<string, boolean> = {};
    const files: Array<string> = [];

    graphWin32.entryPoints.forEach((m) =>
      visit(m, graphWin32, [rootPathWin32], visited, files)
    );
    expect(files).toMatchSnapshot();
  });

  test("returns a list of files including react-native (posix)", () => {
    const visited: Record<string, boolean> = {};
    const files: Array<string> = [];

    graphPosix.entryPoints.forEach((m) =>
      visit(
        m,
        graphPosix,
        [rootPathPosix, react_native_posix.path],
        visited,
        files
      )
    );
    expect(files).toMatchSnapshot();
  });

  test("returns a list of files including react-native (win32)", () => {
    const visited: Record<string, boolean> = {};
    const files: Array<string> = [];

    graphWin32.entryPoints.forEach((m) =>
      visit(
        m,
        graphWin32,
        [rootPathWin32, react_native_win32.path],
        visited,
        files
      )
    );
    expect(files).toMatchSnapshot();
  });
});

describe("TypeScriptValidation()", () => {
  beforeEach(() => {
    yargs.mockReturnValue({
      argv: {
        platform: "test-platform",
      },
    });

    fs.readFileSync.mockReturnValue(
      JSON.stringify({
        include: ["src"],
      })
    );

    child_process.spawnSync.mockReturnValue({
      status: 0,
    });
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  test("replaces file specification with list from graph", () => {
    TypeScriptValidation()(undefined, undefined, graphPosix, {
      projectRoot: rootPathPosix,
    } as SerializerOptions);

    expect(fs.writeFileSync).toBeCalledTimes(1);
    const tsconfig = JSON.parse(fs.writeFileSync.mock.calls[0][1]);
    expect(tsconfig?.include).toBeFalsy();
    expect(tsconfig?.exclude).toBeFalsy();
    expect(tsconfig?.files).toBeTruthy();
    expect(Array.isArray(tsconfig.files)).toBeTruthy();
    expect(tsconfig.files.length).toBeGreaterThan(0);
  });

  test("includes only ts/tsx files when allowJs and checkJs are off", () => {
    TypeScriptValidation()(undefined, undefined, graphPosix, {
      projectRoot: rootPathPosix,
    } as SerializerOptions);

    expect(fs.writeFileSync).toBeCalledTimes(1);
    const tsconfig = JSON.parse(fs.writeFileSync.mock.calls[0][1]);
    expect(tsconfig.files).toMatchSnapshot();
  });

  test("includes ts/tsx and js/jsx files when allowJs and checkJs are on", () => {
    fs.readFileSync.mockReturnValue(
      JSON.stringify({
        include: ["src"],
        compilerOptions: {
          allowJs: true,
          checkJs: true,
        },
      })
    );

    TypeScriptValidation()(undefined, undefined, graphPosix, {
      projectRoot: rootPathPosix,
    } as SerializerOptions);

    expect(fs.writeFileSync).toBeCalledTimes(1);
    const tsconfig = JSON.parse(fs.writeFileSync.mock.calls[0][1]);
    expect(tsconfig.files).toMatchSnapshot();
  });

  test("adds noEmit compiler option", () => {
    TypeScriptValidation()(undefined, undefined, graphPosix, {
      projectRoot: rootPathPosix,
    } as SerializerOptions);

    expect(fs.writeFileSync).toBeCalledTimes(1);
    const tsconfig = JSON.parse(fs.writeFileSync.mock.calls[0][1]);
    expect(tsconfig?.compilerOptions?.noEmit).toBeTruthy();
  });

  function testResolutionPlatforms(
    platform: string,
    expectedPlatforms: string[]
  ) {
    yargs.mockReturnValue({
      argv: {
        platform,
      },
    });

    TypeScriptValidation()(undefined, undefined, graphPosix, {
      projectRoot: rootPathPosix,
    } as SerializerOptions);

    expect(fs.writeFileSync).toBeCalledTimes(1);
    const tsconfig = JSON.parse(fs.writeFileSync.mock.calls[0][1]);
    expect(tsconfig?.compilerOptions?.resolutionPlatforms).toEqual(
      expectedPlatforms
    );
  }

  test("adds list of resolution platforms (android)", () =>
    testResolutionPlatforms("android", ["android", "native"]));

  test("adds list of resolution platforms (ios)", () =>
    testResolutionPlatforms("ios", ["ios", "native"]));

  test("adds list of resolution platforms (macos)", () =>
    testResolutionPlatforms("macos", ["macos", "native"]));

  test("adds list of resolution platforms (windows)", () =>
    testResolutionPlatforms("windows", ["windows", "win", "native"]));

  test("adds list of resolution platforms (win32)", () =>
    testResolutionPlatforms("win32", ["win32", "win", "native"]));

  test("adds list of resolution platforms (win)", () =>
    testResolutionPlatforms("win", ["win", "native"]));

  test("adds list of resolution platforms (FaKe-PLAtfORm)", () =>
    testResolutionPlatforms("FaKe-PLAtfORm", ["fake-platform", "native"]));

  test("runs typescript compiler", () => {
    TypeScriptValidation()(undefined, undefined, graphPosix, {
      projectRoot: rootPathPosix,
    } as SerializerOptions);

    expect(child_process.spawnSync).toBeCalledTimes(1);
    const spawnSyncParams = child_process.spawnSync.mock.calls[0];
    const { [1]: args } = spawnSyncParams;
    expect(args[0]).toEqual(
      expect.stringMatching(
        /[/\\]@msfast[/\\]typescript-platform-resolution.*[/\\]tsc.js$/
      )
    );
  });

  test("preserves extends prop when it refers to a package", () => {
    fs.readFileSync.mockReturnValue(
      JSON.stringify({
        include: ["src"],
        extends: "react-native/tsconfig.json",
      })
    );

    TypeScriptValidation()(undefined, undefined, graphPosix, {
      projectRoot: rootPathPosix,
    } as SerializerOptions);

    expect(fs.writeFileSync).toBeCalledTimes(1);
    const tsconfig = JSON.parse(fs.writeFileSync.mock.calls[0][1]);
    expect(tsconfig.extends).toEqual("react-native/tsconfig.json");
  });

  test("preserves extends prop when it is an absolute path", () => {
    fs.readFileSync.mockReturnValue(
      JSON.stringify({
        include: ["src"],
        extends: "/Users/lithgow/repos/react-native/tsconfig.json",
      })
    );

    TypeScriptValidation()(undefined, undefined, graphPosix, {
      projectRoot: rootPathPosix,
    } as SerializerOptions);

    expect(fs.writeFileSync).toBeCalledTimes(1);
    const tsconfig = JSON.parse(fs.writeFileSync.mock.calls[0][1]);
    expect(tsconfig.extends).toEqual(
      "/Users/lithgow/repos/react-native/tsconfig.json"
    );
  });

  test("re-relativizes extends prop when it is a relative path", () => {
    fs.readFileSync.mockReturnValue(
      JSON.stringify({
        include: ["src"],
        extends: "../../tsconfig.json",
      })
    );

    TypeScriptValidation()(undefined, undefined, graphPosix, {
      projectRoot: rootPathPosix,
    } as SerializerOptions);

    expect(fs.writeFileSync).toBeCalledTimes(1);
    const tsconfig = JSON.parse(fs.writeFileSync.mock.calls[0][1]);
    expect(tsconfig.extends).toEqual(
      expect.stringMatching(/^[.]{2}[/\\][.]{2}[/\\][.]{2}[/\\]tsconfig.json$/)
    );
  });

  test("runs typescript compiler", () => {
    TypeScriptValidation()(undefined, undefined, graphPosix, {
      projectRoot: rootPathPosix,
    } as SerializerOptions);

    expect(child_process.spawnSync).toBeCalledTimes(1);
    const spawnSyncParams = child_process.spawnSync.mock.calls[0];
    const { [1]: args } = spawnSyncParams;
    expect(args[0]).toEqual(
      expect.stringMatching(
        /[/\\]@msfast[/\\]typescript-platform-resolution.*[/\\]tsc.js$/
      )
    );
  });

  test("cleans up temporary tsconfig file when typescript compiler succeeds", () => {
    TypeScriptValidation()(undefined, undefined, graphPosix, {
      projectRoot: rootPathPosix,
    } as SerializerOptions);

    expect(fs.unlinkSync).toBeCalledTimes(1);
    expect(fs.unlinkSync).toBeCalledWith(
      expect.stringMatching(/[/\\]node_modules[/\\]tsconfig-metro.json$/)
    );
  });

  test("cleans up temporary tsconfig file when typescript compiler fails", () => {
    child_process.spawnSync.mockReturnValue({
      status: 1,
    });

    expect(() => {
      TypeScriptValidation()(undefined, undefined, graphPosix, {
        projectRoot: rootPathPosix,
      } as SerializerOptions);
    }).toThrowError();
  });
});
