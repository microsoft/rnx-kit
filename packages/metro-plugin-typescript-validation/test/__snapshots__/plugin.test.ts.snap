// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeScriptValidation() includes only ts/tsx files when allowJs and checkJs are off 1`] = `
Array [
  "/repos/yoyodyne/packages/overthruster/src/main.ts",
  "/repos/yoyodyne/packages/overthruster/src/propulsion.ts",
  "/repos/yoyodyne/packages/overthruster/src/dimensions.ts",
]
`;

exports[`TypeScriptValidation() includes ts/tsx and js/jsx files when allowJs and checkJs are on 1`] = `
Array [
  "/repos/yoyodyne/packages/overthruster/src/main.ts",
  "/repos/yoyodyne/packages/overthruster/src/propulsion.ts",
  "/repos/yoyodyne/packages/overthruster/src/dimensions.ts",
  "/repos/yoyodyne/packages/overthruster/src/lectroid.js",
]
`;

exports[`readTsConfig() reads the contents of tsconfig.json 1`] = `
Object {
  "compilerOptions": Object {
    "noEmit": true,
  },
  "include": Array [
    "src",
  ],
}
`;

exports[`runTypeScriptCompiler() throws when the process crashes 1`] = `"TypeScript validation crashed due to SIGKILL"`;

exports[`runTypeScriptCompiler() throws when the process exit code is non-zero 1`] = `"TypeScript validation failed with exit code 1"`;

exports[`runTypeScriptCompiler() throws when the process exit code is non-zero and an error is given 1`] = `"simulated error message from typescript compiler"`;

exports[`runTypeScriptCompiler() throws when the process exit code is non-zero and an error is given 2`] = `"simulated crash message caused by sending SIGKILL to the typescript compiler"`;

exports[`visit() returns a list of files in the current package (posix) 1`] = `
Array [
  "/repos/yoyodyne/packages/overthruster/src/main.ts",
  "/repos/yoyodyne/packages/overthruster/src/propulsion.ts",
  "/repos/yoyodyne/packages/overthruster/src/dimensions.ts",
  "/repos/yoyodyne/packages/overthruster/src/lectroid.js",
]
`;

exports[`visit() returns a list of files in the current package (win32) 1`] = `
Array [
  "C:\\\\repos\\\\yoyodyne\\\\packages\\\\overthruster\\\\src\\\\main.ts",
  "C:\\\\repos\\\\yoyodyne\\\\packages\\\\overthruster\\\\src\\\\propulsion.ts",
  "C:\\\\repos\\\\yoyodyne\\\\packages\\\\overthruster\\\\src\\\\dimensions.ts",
  "C:\\\\repos\\\\yoyodyne\\\\packages\\\\overthruster\\\\src\\\\lectroid.js",
]
`;

exports[`visit() traverses the entire graph (posix) 1`] = `
Object {
  "/repos/yoyodyne/node_modules/react-native/index.js": true,
  "/repos/yoyodyne/packages/overthruster/src/dimensions.ts": true,
  "/repos/yoyodyne/packages/overthruster/src/lectroid.js": true,
  "/repos/yoyodyne/packages/overthruster/src/main.ts": true,
  "/repos/yoyodyne/packages/overthruster/src/propulsion.ts": true,
}
`;

exports[`visit() traverses the entire graph (win32) 1`] = `
Object {
  "/repos/yoyodyne/packages/overthruster\\\\src\\\\propulsion.ts": true,
  "C:\\\\repos\\\\yoyodyne\\\\node_modules\\\\react-native\\\\index.js": true,
  "C:\\\\repos\\\\yoyodyne\\\\packages\\\\overthruster\\\\src\\\\dimensions.ts": true,
  "C:\\\\repos\\\\yoyodyne\\\\packages\\\\overthruster\\\\src\\\\lectroid.js": true,
  "C:\\\\repos\\\\yoyodyne\\\\packages\\\\overthruster\\\\src\\\\main.ts": true,
  "C:\\\\repos\\\\yoyodyne\\\\packages\\\\overthruster\\\\src\\\\propulsion.ts": true,
}
`;
